File: cpp/dataStructures/basic/unit/testList.cpp
/** \file TestList.cpp
 *
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

#include "Adt.h"
#include "List.h"
#include "Utest.h"

using namespace grafalgo;

bool basicTests() {
	List l1; int n1 = l1.n();

	Utest::assertTrue(l1.empty(), "initial list not empty");
	Utest::assertTrue(l1.isConsistent(), "initial list not consistent");
	cout << "writing empty list: " << l1 << endl;

	for (int i = 1; i <= n1; i++) {
		Utest::assertTrue(!l1.member(i),
			"member returns true on empty list");
	}

	l1.addLast(1);
	cout << "writing one item list: " << l1 << endl;
	Utest::assertEqual(l1.toString(), "[a]",
		"mismatch on adding first item");
	Utest::assertTrue(l1.isConsistent(), "new list not consistent");
	
	Utest::assertTrue(!l1.empty(),"list with one item reports empty");
	Utest::assertTrue(l1.member(1),"member 1 not reported in list");

	for (int i = 3; i <= n1; i += 2) l1.addLast(i);
	cout << "writing longer list: " << l1 << endl;
	Utest::assertEqual(l1.toString(), "[a c e g i]",
		"mismatch on list [a c e g i]");
	Utest::assertTrue(l1.isConsistent(), "longer list not consistent");

	Utest::assertTrue(l1.member(5),
		"member 5 not reported in list");
	Utest::assertTrue(!l1.member(4),
		"non-member 4 is reported in list");

	l1.removeFirst();
	Utest::assertEqual(l1.toString(), "[c e g i]",
		"mismatch on list [c e g i]");
	Utest::assertTrue(l1.isConsistent(), "not consistent after removal");
	Utest::assertTrue(!l1.member(1),
		"non-member 1 is reported in list");

	l1.removeNext(5); 
	Utest::assertEqual(l1.toString(), "[c e i]",
		"mismatch on list [c e i]");

	l1.removeNext(0);
	Utest::assertEqual(l1.toString(), "[e i]",
		"mismatch on list [e i]");
	Utest::assertTrue(l1.isConsistent(),
			  "not consistent after next removal");

	l1.removeFirst();
	Utest::assertEqual(l1.toString(), "[i]",
		"mismatch on list [i]");
	Utest::assertTrue(l1.isConsistent(),
			  "not consistent after next removal");

	Utest::assertTrue(!l1.empty(),
		"non-empty list reported as empty");

	l1.removeFirst();
	Utest::assertEqual(l1.toString(), "[]",
		"mismatch on list []");
	Utest::assertTrue(l1.isConsistent(),
			  "not consistent after final removal");

	Utest::assertTrue(l1.empty(),
		"empty list reported as non-empty");

	l1.addFirst(1); l1.addFirst(2); l1.addFirst(3);
	List l2; l2 = l1;
	Utest::assertEqual(l2.toString(), "[c b a]",
		"mismatch on list [c b a]");
	int n2 = 27; l2.expand(n2);
	Utest::assertEqual(l2.n(),27,"mismatch on size of expanded set");
	Utest::assertEqual(l2.toString(), "[3 2 1]",
		"mismatch on list [3 2 1]");

	l2.addLast(30);
	Utest::assertTrue(l2.member(30),"mismatch on membership test (30)");
	Utest::assertTrue(!l2.member(29),"mismatch on membership test (29)");
	Utest::assertEqual(l2.n(),54,"mismatch on size of expanded set (54)");

	for (int i = 31; i <= 60; i++) l2.addLast(i);
	for (int i = 31; i <= 60; i++)
		Utest::assertTrue(l2.member(i),
			"mismatch on membership test (" + to_string(i) + ")");
	Utest::assertEqual(l2.n(),108,"mismatch on size of expanded set (108)");

	l2.resize(30); l2.addFirst(1); l2.addFirst(2); l2.addFirst(3);
	cout << "writing numeric list: " << l2 << endl;
	Utest::assertEqual(l2.toString(), "[3 2 1]",
		"mismatch on list [3 2 1]");
	Utest::assertTrue(l1.isConsistent(),
			  "not consistent after three more adds");

	// test equality operator
	l1.clear(); l2.clear();
	Utest::assertTrue(l1 == l1,"==: list testing unequal "
					"to itself");
	Utest::assertTrue(l1 == l2,"==: empty lists testing as "
					"unequal");
	l1.addFirst(1);
	Utest::assertTrue(!(l1 == l2),"==: different lists testing as "
					"equal");
	l2.addLast(1);
	Utest::assertTrue(l1 == l2,"==: equal lists testing as "
					"unequal");
	l1.addLast(5); l2.addLast(5);
	l1.addFirst(3); l2.addFirst(3);
	Utest::assertTrue(l1 == l2,"==: equal lists testing as "
					"unequal");
	l1.removeFirst();
	Utest::assertTrue(!(l1 == l2),"==: unequal lists testing as "
					"equal");

	return true;
}

/**
 *  Unit test for List data structure.
 */
int main() {
	cout << "running basic tests\n";
	if (basicTests()) cout << "basic tests passed\n";

	// add more systematic tests for each individual method
}

File: java/dataStructures/basic/unit/testList.java
/** \file TestList.java
 *
 *  @author Jon Turner
 *  @date 2021
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import java.util.Scanner;
import java.util.ArrayList;
import grafalgo.misc.Util;
import grafalgo.dataStructures.Adt;
import grafalgo.dataStructures.basic.List;

public final class testList {
	public static boolean basicTests() {
		int n1 = 10; List l1 = new List(n1);

		l1.assertEqual(l1.n(), 10, "a1");
	
		l1.assertTrue(l1.empty(), "a2");
		l1.assertConsistent("a3");
	
		for (int i = 1; i <= n1; i++)
			l1.assertTrue(!l1.member(i), "a4 " + i);
	
		l1.insert(1, 0); l1.assertEqual("[a]", "b1");
		l1.assertConsistent("b2");
		l1.assertTrue(!l1.empty(),"b3");
		l1.assertTrue(l1.member(1),"b4");
	
		for (int i = 3; i <= n1; i += 2) l1.enq(i);
		l1.assertEqual("[a c e g i]", "c1");
		l1.assertConsistent("c2");
		l1.assertTrue(l1.member(5), "c3");
		l1.assertTrue(!l1.member(4), "c4");
	
		int v = l1.pop(); l1.assertEqual("[c e g i]", v, 1, "d1");
		l1.assertConsistent("d2");
		l1.assertTrue(!l1.member(1), "d3");
		l1.insert(8, 5); l1.assertEqual("[c e h g i]", "d4");
		l1.deleteNext(8); l1.assertEqual("[c e h i]", "d5");
		l1.deleteNext(8); l1.assertEqual("[c e h]", "d6");
		l1.deleteNext(0); l1.assertEqual("[e h]", "d7");
		l1.push(4); l1.assertEqual("[d e h]", "d8");
		l1.assertConsistent("d9");
		l1.assertTrue(!l1.empty(), "d10");
	
		int n2 = 27; List l2 = new List(n2);
		l2.push(1); l2.push(2); l2.push(3);
		l2.assertEqual("[3 2 1]", "e1");
		l1.assertConsistent("e2");
		l1.clear(); l1.enq(3); l1.enq(2); l1.enq(1);
		l1.assertEqual(l2,"e3");
		l1.assertEqual("[c b a]", "e4");

		Scanner in = new Scanner("[b c d e]");
		l1.read(in); l1.assertEqual("[b c d e]", "f1");
		l1.expand(15); l1.assertEqual(l1.n(), 15, "f2");
		in = new Scanner("[ b  g   27 ]");
		l1.read(in); l1.assertEqual("[2 7 27]", l1.n(), 27, "f3");
		l1.reset(10); l1.assertEqual("[]", "f4");
		
		return true;
	}
	
	/**
	 *  Unit test for List data structure.
	 */
	public static void main(String[] args) {
		System.out.println("running basic tests");
		if (basicTests()) System.out.println("basic tests passed");
	
		// add more systematic tests for each individual method
	}
}
